### PR TITLE
CircleCI

### PR DESCRIPTION
**Pull request by shen-tian**
Part of a shift of CI from Travis to CircleCI. This is very similar to nrepl/piggieback#102

See this for what test runs look like:

[CircleCI](https://circleci.com/gh/shen-tian/nrepl/tree/circleci)

There's some issues around tests randomly failing from time to time. This also happened on Travis, but I haven't been able to reproduce locally yet.

### CODE REVIEW

**Comment by shen-tian** [comment]
Examples of failures:

Java 8, Clojure 1.10 (https://circleci.com/gh/shen-tian/nrepl/4#tests/containers/0)

```
lein test :only nrepl.core-test/streamed-printing

ERROR in (streamed-printing) (bencode.clj:177)
bencode transport

Uncaught exception, not in assertion.
expected: nil
  actual: java.lang.ClassCastException: clojure.lang.PersistentArrayMap cannot be cast to [B
 at nrepl.bencode$string_LT_payload.invokeStatic (bencode.clj:177)
    nrepl.bencode$string_LT_payload.invoke (bencode.clj:177)
    nrepl.bencode$read_map$fn__375.invoke (bencode.clj:291)
    clojure.core$map$fn__5847$fn__5848.invoke (core.clj:2742)
```
Java8 Clojure 1.7 (https://circleci.com/gh/shen-tian/nrepl/20)

```
lein test nrepl.core-test

lein test :only nrepl.core-test/streamed-printing

ERROR in (streamed-printing) (bencode.clj:177)
bencode transport

Uncaught exception, not in assertion.
expected: nil
  actual: java.lang.ClassCastException: clojure.lang.PersistentArrayMap cannot be cast to [B
 at nrepl.bencode$string_LT_payload.invoke (bencode.clj:177)
    nrepl.bencode$read_map$fn__222.invoke (bencode.clj:291)
    clojure.core$map$fn__4549$fn__4550.invoke (core.clj:2611)
    clojure.core$partition_all$fn__6811$fn__6812.invoke (core.clj:6971)
    clojure.core.protocols/fn (protocols.clj:167)
    clojure.core.protocols$fn__6478$G__6473__6487.invoke (protocols.clj:19)
    clojure.core.protocols$seq_reduce.invoke (protocols.clj:31)
 ```

Java11 Clojure 1.9 (https://circleci.com/gh/shen-tian/nrepl/25)

```
lein test :only nrepl.core-test/streamed-printing

ERROR in (streamed-printing) (bencode.clj:177)
bencode transport

Uncaught exception, not in assertion.
expected: nil
  actual: java.lang.ClassCastException: class clojure.lang.PersistentArrayMap cannot be cast to class [B (clojure.lang.PersistentArrayMap is in unnamed module of loader 'app'; [B is in module java.base of loader 'bootstrap')
 at nrepl.bencode$string_LT_payload.invokeStatic (bencode.clj:177)
    nrepl.bencode$string_LT_payload.invoke (bencode.clj:177)
    nrepl.bencode$read_map$fn__384.invoke (bencode.clj:291)
```

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  working_directory: ~/repo
+  environment:
+    LEIN_ROOT: "true"
+    JVM_OPTS: -Xmx3200m
+
+
+executors:
+  openjdk8:
+    docker:
+      - image: circleci/clojure:openjdk-8-lein-2.9.0
```
Comment (bbatsov) [inline .circleci/config.yml]: There 2.9.1 now.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,188 @@
+version: 2.1
+
+defaults: &defaults
+  working_directory: ~/repo
+  environment:
+    LEIN_ROOT: "true"
+    JVM_OPTS: -Xmx3200m
```
Comment (bbatsov) [inline .circleci/config.yml]: Where you getting some errors without this?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+          command: |
+            sudo apt-get install make
+      - run:
+          name: Running tests
+          command: make test
+
+  codecov:
+    executor: openjdk11
+    environment:
+      VERSION: "1.9"
```
Comment (bbatsov) [inline .circleci/config.yml]: Why 1.9?

File: .circleci/config.yml (hunk 2 above)
Comment (shen-tian) [reply .circleci/config.yml]: No. Just a carry over from another config. Can try without.

File: .circleci/config.yml (hunk 3 above)
Comment (shen-tian) [reply .circleci/config.yml]: No reason in particular. Should we do the `utility` tasks on JDK11/Clojure 1.9?

File: .circleci/config.yml (hunk 2 above)
Comment (bbatsov) [reply .circleci/config.yml]: Got it. I noticed this in the example provided by Circle themselves, so I guess that's how it became popular. I'm fine with it, I was just curious how this number was devised.

File: .circleci/config.yml (hunk 3 above)
Comment (bbatsov) [reply .circleci/config.yml]: Probably they should be on Java 11 and Clojure 1.10.

**Comment by bbatsov** [comment]
As mentioned on another ticket, I really like this a structure https://github.com/clojure-emacs/cider/blob/master/.circleci/config.yml (it was authored by @danielcompton recently).

**Comment by bbatsov** [comment]
(but for Clojure we don't need to run the lint tasks on every build)

File: .circleci/config.yml (hunk 2 above)
Comment (danielcompton) [reply .circleci/config.yml]: It's important to set `JVM_OPTS` when running in CircleCI. Your jobs run on massive machines with tons of RAM, but inside a container (or similar) with a much smaller allocation of memory. By default older JVMs will set Xmx to a value higher than it is actually allowed to allocate. On quick builds, you may not generate enough garbage to hit the limit, or may only sometimes hit it. Newer JVM's (9 and 10 have both had improvements here) are smart enough to detect their effective max memory and set it by default, but JDK 8 doesn't have this IIRC.

**Comment by shen-tian** [comment]
Jumping back here from nrepl/piggieback#102 and clojure-emacs/cider-nrepl#602. Let's get this one done first, and the others should follow?

I would like to get some input on the "how many jobs do we need" question. Discussion [here](https://github.com/nrepl/piggieback/pull/102#issuecomment-471420100).

**Comment by bbatsov** [comment]
> Jumping back here from nrepl/piggieback#102 and clojure-emacs/cider-nrepl#602. Let's get this one done first, and the others should follow?

Yeah, I agree.

> I would like to get some input on the "how many jobs do we need" question. Discussion here.

Unfortunately I can't provide much input myself, as I started using Circle relatively recently. It's fine not to get things perfect from the start.

One thing that I can suggest is to annotate the build file with some comments highlighting the design decisions there. That would make it easier to tweak things down the line I think, at least for someone new to Circle, like me.

File: .circleci/config.yml (hunk 2 above)
Comment (shen-tian) [reply .circleci/config.yml]: Good to know!

**Comment by bbatsov** [comment]
@shen-tian I love the current version! Great work!

Seems the only thing left is to remove the Travis stuff and update the badge in the README.

**Comment by shen-tian** [comment]
I'm quite happy with where the config is sitting at now. Thanks for all the input/suggestions! 

- It runs as a group of independent jobs, one for each cell in test matrix, and three utility jobs. There isn't any dependency on a `checkout_code` job. This makes quite easy to see which job is passing/failing straight from Github, and everything takes less than 3min to run.
- Utility jobs all run on Java 11 and Clojure 1.10
- The config file is reasonably well annotated, and most of it could probably be shared via [Orbs](https://circleci.com/orbs/), a new feature that makes easy to share CI jobs etc. between projects. I suspect we'll find a lot in common between the cider/nrepl Clojure projects.

What's not great... is that the tests aren't reliably passing. It's the same failure as I posted above, but the same failure mode is also happening in Travis, so I think that's a "make tests more robust" job, rather than a configure CI better one?

**Comment by bbatsov** [comment]
> What's not great... is that the tests aren't reliably passing. It's the same failure as I posted above, but the same failure mode is also happening in Travis, so I think that's a "make tests more robust" job, rather than a configure CI better one?

Yeah, I completely agree. As far as I'm concerned the build config is good to go.

**Comment by shen-tian** [comment]
> @shen-tian I love the current version! Great work!
> 
> Seems the only thing left is to remove the Travis stuff and update the badge in the README.

Think you need to

**Pull request merged**